File: tools/ts/lcm/examples/publisher.ts
#!/usr/bin/env -S deno run --allow-net --unstable-net

// Example: Publishing LCM messages
// Run with: deno task pub

import { LCM } from "@dimos/lcm";
import { Point } from "@dimos/msgs/geometry_msgs/mod.ts";

async function main() {
  const lcm = new LCM();
  await lcm.start();

  console.log("Publishing Point messages on 'EXAMPLE_POINT'...");
  console.log("Press Ctrl+C to stop.\n");

  let counter = 0;

  while (lcm.isRunning()) {
    const point = new Point({
      x: Math.sin(counter * 0.1) * 10,
      y: Math.cos(counter * 0.1) * 10,
      z: counter * 0.1,
    });

    await lcm.publish("EXAMPLE_POINT", point);
    console.log(`Published: x=${point.x.toFixed(2)}, y=${point.y.toFixed(2)}, z=${point.z.toFixed(2)}`);

    counter++;
    await new Promise((r) => setTimeout(r, 100)); // 10 Hz
  }
}

main().catch(console.error);

File: tools/rust/lcm/examples/publisher.rs
// Simple LCM Publisher Example
// Publishes Vector3 messages at 10 Hz

use dimos_lcm::LcmModule;
use lcm_msgs::geometry_msgs::Vector3;
use tokio::time::{sleep, Duration, Instant};

#[tokio::main]
async fn main() {
    let mut publisher = LcmModule::new().await.expect("Failed to create LcmModule (publisher)");

    // in a Native Module, this will be handled by parsing the args
    publisher.map_topic("vector", "/vector#geometry_msgs.Vector3");

    let vector = publisher.output("vector", Vector3::encode);
    let _handle = publisher.spawn();

    println!("Publishing Vector3 on {}", vector.topic);
    println!("Press Ctrl+C to stop\n");

    let mut t: f64 = 0.0;
    let mut last = Instant::now();

    loop {
        let vec = Vector3 {
            x: t.sin() * 5.0,
            y: t.cos() * 5.0,
            z: t,
        };

        let interval = last.elapsed();
        last = Instant::now();

        let _ = vector.publish(&vec).await;

        println!(
            "Published: x={:.2} y={:.2} z={:.2} (interval {:.1}ms)",
            vec.x, vec.y, vec.z, interval.as_secs_f64() * 1000.0
        );

        t += 0.1;
        sleep(Duration::from_millis(100)).await;
    }
}
